content/news/2019-12-03-mapstruct-and-quarkus.md

```diff
@@ -0,0 +1,220 @@
+
+{{< prettify >}}
+mvn io.quarkus:quarkus-maven-plugin:1.0.1.Final:create \
+  -DprojectGroupId=org.mapstruct.examples.quarkus \
+  -DprojectArtifactId=mapstruct-examples-quarkus \
+  -DclassName="org.mapstruct.example.quarkus.PersonResource" \
+  -Dpath="/person" \
+  -Dextensions="resteasy-jsonb"
+{{< /prettify >}}
+
+This will setup a new Quarkus application having a REST endpoint `/person` defined in a `PersonResource`. We want to expose the JSON structure automatically, so we need the `resteasy-jsonb` extension that will allow us to marshall our object to JSON (like you probably know it from Spring REST or similar frameworks).
+
+#### Adding MapStruct
+You just need to add MapStruct in your projects POM:
+
+{{< prettify xml >}}
+<dependency>
+  <groupId>org.mapstruct</groupId>
+  <artifactId>mapstruct</artifactId>
+  <version>1.3.1.Final</version>
+</dependency>
+<dependency>
+  <groupId>org.mapstruct</groupId>
+  <artifactId>mapstruct-processor</artifactId>
+  <version>1.3.1.Final</version>
+  <scope>provided</scope>
+</dependency>
+{{< /prettify >}}
+
+_Side note:_ In case you use for example Lombok pay attention that you define the dependency **before** the one for the MapStruct processor.
+
+#### Person service
+This example application will have a `Person` POJO holding information about the person and a `PersonService` that will return the person.
+
+So we have to create the following two classes:
+
+{{< prettify java >}}
+public class Person {
+  private String firstname;
+  private String lastname;
+  
+  // all-args constructor, getters, setters
+}
+{{< /prettify >}}
+
+{{< prettify java >}}
+@ApplicationScoped
+public class PersonService {
+  public Person loadPerson() {
+    return new Person("Bob", "Miller");
+  }
+}
+{{< /prettify >}}
+
+What is the first thing you notice?
+
+Okay… the service is very boring. It always returns the same object. But that’s not what I meant, it’s just an example.. Most likely, your application will fetch the information from an external system.
+
+What I meant is the `@ApplicationScoped` annotation that was used. If you are just a Spring guy you might not know it, but this is a regular Java EE annotation that belongs to [CDI](https://www.jcp.org/en/jsr/detail?id=299) and marks this class as (more or less) an application-wide singleton that is “injectable”. Similar to `@Component` or `@Service` from Spring.
+
+Most of the commonly used frameworks are using reflection for the DI and as mentioned before reflection doesn’t play well with native images. But the Quarkus team did a great job and provided a CDI implementation that can be used together with native images - so that’s then most likely the way to go when you need dependency injection and build a Quarkus application!
+
+#### DTO and mapper
+We don’t like to expose internal objects, thus we create a new `PersonDto` that matches our API.
+
+{{< prettify java >}}
+public class PersonDto {
+  private String firstname;
```

Comment:
Perhaps make them public? I think it's a good practice for DTOs.

Reply:
I personally prefer the getter/setter style _(+ lombok :grimacing:)_ so I will leave it as it is